### DISCUSSION TITLE
Proxy in Docker Container on a Raspberry Pi 4. How to run it?

### DISCUSSION THREAD

**Discussion by WadjoResende** [s-allius/tsun-gen3-proxy#142] [Q&A] [answered]
Hello!
Congratulations for the proxy!
I'm trying to run tsun-gen3-proxy in a docker container on a raspberry pi 4, but it doesn't work...
In the log view I receive:

"exec /root/entrypoint.sh: exec format error
exec /root/entrypoint.sh: exec format error
exec /root/entrypoint.sh: exec format error
exec /root/entrypoint.sh: exec format error
exec /root/entrypoint.sh: exec format error
exec /root/entrypoint.sh: exec format error
..."

My docker compose is:

  tsun-proxy:
    container_name: tsun-proxy
    image: ghcr.io/s-allius/tsun-gen3-proxy:latest
    # image: ghcr.io/s-allius/tsun-gen3-proxy:rc
    restart: unless-stopped
    depends_on:
    - mosquitto
    environment:
    - TZ=America/Sao_Paulo
    - PUID=1000
    - PGID=1000
    ports: 
    - 5005:5005
    - 8127:8127
    - 10000:10000
    volumes:
    - ./volumes/tsun/log:/home/tsun-proxy/log 
    - ./volumes/tsun/config:/home/tsun-proxy/config
    
    Could anybody help me with it?
Thanks!

**Comment by s-allius** [comment]
Hi,

I think the problem is, that a raspberry is an ARM architecture and I only build the Intel packages and uploaded these.
You must build the container on your target.
Please try to build and run the container on your target system, like I have described here: https://github.com/s-allius/tsun-gen3-proxy?tab=readme-ov-file#getting-started

If this work I can add a package for ARM to the packages store.

Cheers Stefan

**Reply by WadjoResende** [reply]
I will try to build the image and as soon as I have some results I will come back here to comment.
Thank you very much for your attention!
Wanderson

**Reply by s-allius** [reply]
I just started to build a multi arch build. But it is still running and needs a long time on my laptop. So wenn this runs well, I can offer a package for testing...

**Reply by s-allius** [reply, marked as answer]
You may try this dev package from docker.io
```
    image: docker.io/sallius/tsun-gen3-proxy:dev
```
It's my first multi arch image

**Reply by WadjoResende** [reply]
Wow!
When I was going to try to build the image, you came up with the solution.
It worked perfectly!
I have two TSOL-MP3000. They are disconnect now (they disconnect automatic at night). I'll try to configure them tomorrow.
Thank you very much!
![proxy](https://github.com/user-attachments/assets/aa36da30-e494-41fe-8773-af01da19f2bd)

**Reply by s-allius** [reply]
You are welcome!
I will adapt my build pipeline, to support ARM architecture by default from the next release.

I think you are the first owner of a TSOL-MP3000 who use the proxy. If you have any questions or need help, please ask and I will try to help you. I'm really interested to support more TSUN inverters.

**Reply by WadjoResende** [reply]
Hi Stefan!
I tried to connect today and ended up losing local access via wifi and also the connection to the cloud. The generation is working normally, but the inverters connections are offline. Now it's night here and the inverters turn off AP access. I think I'll only be able to try to solve the problem on Saturday, which is when I'll be home during the day.
As soon as I have some results with the connection, I will let you know.

**Reply by s-allius** [reply]
If you have lost the Wifi connection, you can reset the TSUN inverter by disconnecting the inverter from the grid over night. So the inverter is completely without power (from grid and PV). The next day you can normally make a new Wifi configuration (Pairing).

That will do with the GEN3 and GEN3PLUS inverter. I hope that this will also work on your TiTAN inverter.

Does your inverter has a WEB interface and broadcast his own WiFi SSID like the GEN3PLUS inverters (e.g. MS2000)?

**Reply by WadjoResende** [reply]
Yes, the TSOL-MP3000 has a web interface that can be accessed via the LAN IP or via AP mode connection. When I tried to configure the inverters, I lost access to them. They are connecting to my router via Wi-Fi, but the web interface does not open via their IP on my network. They still accept the connection via AP mode, but the interface does not open in this mode either.
This morning and late afternoon I tried every way to connect to the inverters, but it was not possible.
It is now 9:30 pm here. I turned off the power to the inverters and I will turn it on again tomorrow morning.
When I get access again, I will send a print of the interface screens here.
I must have done something wrong on the first attempt and would like your help to configure this connection. I believe I will need to connect according to your tutorial "Client Mode (GEN3PLUS only)", since my inverters were connected to the cloud via port 10443.
Thank you very much for your attention!

**Reply by WadjoResende** [reply]
This morning I managed to recover the connection of one of them. The other one did not open the interface.
The web interface of the inverter that recovered the connection to the cloud is no longer opening in my browser. This is strange... I don't understand how this connection works. It seems that the inverter blocks access for some reason that I don't know.
Tonight I will turn off the inverters again to try to recover the connection of the other inverter tomorrow.

**Reply by s-allius** [reply]
I had a similar problem with my MS2000. I must additional deactivate the used SSID during the start. Then I could pair the inverter with the APP again and the TCP connection worked again. 
I think when the inverter can't connect to an WiFi network after a power cycle, he resets something internaly. 

Sometimes I had the feeling, that waiting is the solution. After a few hours the web interface was working again...

**Reply by WadjoResende** [reply]
Hello Stefan!
Today I re-established the connection of the inverters to the TSUN cloud and made another attempt to configure them in the proxy.
Below are some screens of the TSOL-MP3000 configuration:

![Imagem do WhatsApp de 2024-08-11 à(s) 09 17 51_ae0cbf96](https://github.com/user-attachments/assets/b833a4b9-bd6e-434b-8e9f-84af7ecd3552)

![captura 192 168 0 230](https://github.com/user-attachments/assets/c3ffebba-a4f6-4423-a4cd-9dab686188a4)

Since the configuration made in the URL http://<ip-adress>/config_hide.html caused the inverters to be completely disconnected, this time I configured it through the URL http://<ip-adress>/index_cn.html, as shown in the print below:

![Configuração inversor 1](https://github.com/user-attachments/assets/4539c1d0-43b8-4ca9-9966-eede3ecc7755)

Apparently the connection was established:

![Log container](https://github.com/user-attachments/assets/f9845a2a-deb0-4c52-a97b-6f08a7e53d08)

Home Assistant reports the connection with two inverters, but no readings are loaded:

![Tela home assistant 1](https://github.com/user-attachments/assets/0021d650-3fe8-4116-b92e-6588f6d7a9af)

How can I send you the "trace.log"?

**Reply by s-allius** [reply]
Ok, let's take a look.
1. your firmware supports SSL. This mean that you cannot connect to Port 10000. You should not configure something on the index_cn side as `Server B`
2. You have configured the `client-mode` and the are connection established. But we see a timeout. Can have a look in the proxy.log file? you will find it on you server in the log directory of you docker container.
There I would expect a heartbeat every 120 sec. But maybe you only get MODBUS timeout.
3. you can send me the log per eMail to Kontakt@allius.de. I assume, that you inverter has a different register mapping and we have to determine them first

**Reply by WadjoResende** [reply]
Nice!
I've sent the logs to your e-mail.
Thank you!

**Reply by s-allius** [reply]
I got your eMail.
Cool, your inverters are sending a heartbeat request and the proxy answers them, properly. That' really good, because, the underlaying protocol is working.
Then the proxy sends MODBUS requests frequently, and your inverter answers them as well. Unfortunately I have deactivated the packet trace for the MODBUS packets in the release proxy, because of the huge log files.
For the next steps I need more informations from the packet trace. Maybe we got an MODBUS error (which I never have seen from an TSUN inverter) or I make an mistake by parsing the response...
Could you please try a debug version of the proxy? You can easily install it by changing the URL in you `docker-compose.yaml` file:
```
    image: docker.io/sallius/tsun-gen3-proxy:debug
``` 
Then send me the new traces again, please.

**Reply by WadjoResende** [reply]
I sent you new logs.
Thanks!

**Reply by s-allius** [reply]
I got your traces and open an issue for supporting the TITAN inverter. => [issue 161](https://github.com/s-allius/tsun-gen3-proxy/issues/161)

My first idea is, that your inverters have a different sensor type. So I build a new debug firmware with a different one, which I guess that it should match

Can you please install the new debug image. (same URL) and send me the traces again?

**Reply by s-allius** [reply]
I tried the debug version with the changed sensor type on a GEN3PLUS inverter and hoped that I would then see the same error. Unfortunately not. I therefore assume that the sensor type is ignored.
I then increased the timeout to rule out a collision due to too fast repetition.
 
The last version was build at:
2024-08-12T17:54+02:00 => Version: 0.10.1+debug
so pls check that you test with the latest one.

**Reply by WadjoResende** [reply]
I've sent you new logs with update from 13:14...

**Reply by s-allius** [reply]
Ok, this don't fix it. The changing of the timeout is good, so we don't get MODBUS timeouts anymore.
But we don't get any data.
So I have to think a little bit about that. Maybe the inverter id is wrong, or the register addresses are in a wrong range. So I can build a proxy which scan the hole register range. For this I need a little bit more time.
Before I start with this I will try to get the response code from my inverter.
I come back to you, if I have a new proxy version for testing...

**Reply by WadjoResende** [reply]
OK. I'll be here for whatever you need.
Thank you!

**Reply by s-allius** [reply]
I can reproduce your error here. Wenn I use an invalid MODUS register or an invalid MODBUS inverter ID, I receive the same response with the 0x0500.
So until you can't have multiple MODUS client on TCP, changing the inverter ID to anything else then 1, makes no sense. So I will start with testing the register range.
I have to write a little bit of code and test it here. I send you a new version in a few hours...

**Reply by s-allius** [reply]
I have a simple scan function implemented. Pls use the dev firmware because we don't won't the huge logs for the scan: 
```
    image: docker.io/sallius/tsun-gen3-proxy:dev
```
In the docker console you will see something like that
```'2024-08-12 20:10:10  INFO | root | Valid MODBUS data (reg: 0x2000):'```
if we hit a valid register range.

to check if the scan runs, look into the proxy.log file. You should see MODBUS timeouts with increasing regs:
'2024-08-12 20:18:28  INFO | data | [] Modbus timeout (FCode: 3 Reg: 0x0100, 48)'
'2024-08-12 20:18:58  INFO | data | [] Modbus timeout (FCode: 3 Reg: 0x0200, 48)'

The scan will run over about 2 hours and 8 minutes.

**Reply by WadjoResende** [reply]
I sent you the logs by email.
Unfortunately, nothing like "Valid MODBUS data"...

Sorry, I don't have enough knowledge to give an opinion on a possible solution.
This is the print of the current inverter communication configuration:
 
![image](https://github.com/user-attachments/assets/662d990d-1d04-42d7-9410-b4b61f23035f)

**Comment by WadjoResende** [comment]
These are the options of protocol:

![image](https://github.com/user-attachments/assets/1f2a7038-01ed-4363-b0c1-482af2fc8165)

**Reply by s-allius** [reply]
What will happen, when you set `TCP Client`and as Port: 10000?

**Reply by WadjoResende** [reply]
I'll try it tomorrow.

**Reply by WadjoResende** [reply]
I tried this setting but it didn't work.
